File: client/src/features/dossier/reviewRules.ts
/**
 * Review state machine — CLIENT-SIDE MIRROR of the server's
 * app/services/review_rules.py. The server always re-validates; this copy
 * exists purely to hint the UI (which buttons to offer, when a note is due).
 */
import type { Disposition } from '@/services/api/v2Types';

export const ALLOWED_TRANSITIONS: Record<Disposition, Disposition[]> = {
  needs_review: ["reviewed", "escalated", "dismissed"],
  reviewed: ["needs_review", "escalated", "dismissed"],
  escalated: ["needs_review"],
  dismissed: ["needs_review"],
};

export const NOTE_REQUIRED: readonly Disposition[] = ["escalated", "dismissed"];
export const MIN_NOTE_CHARS = 10;

export function allowedActions(status: Disposition): Disposition[] {
  return ALLOWED_TRANSITIONS[status] ?? [];
}

export function isNoteRequired(action: Disposition): boolean {
  return NOTE_REQUIRED.includes(action);
}

export function isNoteSufficient(note: string): boolean {
  return note.trim().length >= MIN_NOTE_CHARS;
}

export const ACTION_LABELS: Record<Disposition, string> = {
  needs_review: "Reopen (needs review)",
  reviewed: "Mark reviewed",
  escalated: "Escalate",
  dismissed: "Dismiss",
};

/** Generate a fresh idempotency key — one per intended submission. */
export function newIdempotencyKey(): string {
  return `rv-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
}

File: server/api/app/services/review_rules.py
"""Analyst review state machine (Phase 4) — the single server-side source of
truth for which disposition changes are legal.

``ALLOWED_TRANSITIONS`` is mirrored by the ReviewControls UI (client-side
hint only) — the server always re-validates before persisting. Review events
are append-only, so an illegal transition attempt is rejected with 422 and
never recorded.
"""
from __future__ import annotations

ALLOWED_TRANSITIONS: dict[str, set[str]] = {
    "needs_review": {"reviewed", "escalated", "dismissed"},
    "reviewed": {"needs_review", "escalated", "dismissed"},
    "escalated": {"needs_review"},
    "dismissed": {"needs_review"},
}

# A substantive rationale is mandatory before escalating or dismissing —
# the audit trail must record WHY a source was written off or promoted.
NOTE_REQUIRED_FOR: frozenset[str] = frozenset({"escalated", "dismissed"})
MIN_NOTE_CHARS = 10


def allowed_actions(status: str) -> list[str]:
    """Sorted list of dispositions reachable from ``status`` (UI hint + 422 detail)."""
    return sorted(ALLOWED_TRANSITIONS.get(status, set()))


def is_transition_valid(current: str, target: str) -> bool:
    """True when ``current -> target`` is an allowed disposition change."""
    return target in ALLOWED_TRANSITIONS.get(current, set())


def note_required(action: str) -> bool:
    """True when ``action`` demands a substantive rationale note."""
    return action in NOTE_REQUIRED_FOR


def note_sufficient(note: str | None) -> bool:
    """True when the note exists and meets the minimum substantive length."""
    return note is not None and len(note.strip()) >= MIN_NOTE_CHARS
